### PR TITLE
Adds Chat Message History and Connects to the NPC Chat API

### PR DESCRIPTION
**Pull request by igooch**
**What type of PR is this?**
/kind documentation

**What this PR does / Why we need it**:
Adds chat message history to the GenAI Request. This will make for better responses from VertexAI.

Adds the request type NPC which allows for making requests to the [Google for Games GenAI NPC API](https://github.com/googleforgames/GenAI-quickstart/genai/api/npc_chat_api).

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:
This works with Vertex, so in https://github.com/googleforgames/GenAI-quickstart/blob/205ba78aa65782f00a8654e31c88efdc4fdaded6/genai/api/npc_chat_api/config/config.toml#L20-L21 comment out "GKEGenAI", and uncomment "VertexAI".

You will need to first start up and run the GenAI Quickstart cluster, then start and run the NPC chat server.

### CODE REVIEW

File: examples/simple-genai-server/gameserver_npcchat.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  template:
+    spec:
+      containers:
+        - name: simple-genai-game-server
+          image: us-docker.pkg.dev/agones-images/examples/simple-genai-game-server:0.1
+          # imagePullPolicy: Always  # add for development
+          env:
+            - name: GenAiEndpoint
+            # Use the service endpoint address when running in the same cluster as the inference server.
+              value: "http://npc-chat-api.genai.svc.cluster.local:80"
```
Comment (zmerlynn) [inline examples/simple-genai-server/gameserver_npcchat.yaml:34]: This should be plumbed to the npc chat endpoint of `/genai` eventually. The idea is for that to be the front-end to all of it. You can add here:

```
TODO: Change this to the `/genai/npc-chat` endpoint when it's properly plumbed. 
```

File: examples/simple-genai-server/gameserver_npcchat.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            # GenAiContext is not passed to the npc-chat-api endpoint.
+            - name: GenAiNpc # False by default. Use GenAiNpc "true" when using the npc-chat-api as the GenAiEndpoint.
+              value: "true"
+            - name: SimEndpoint
+              value: "http://vertex-chat-api.genai.svc.cluster.local:80"
```
Comment (zmerlynn) [inline examples/simple-genai-server/gameserver_npcchat.yaml:43]: Here use the `/genai/chat` endpoint instead. If you want to handle this as a follow-on and leave a TODO that's fine, though.

File: examples/simple-genai-server/gameserver_npcchat.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+              value: "50"
+          resources:
+            requests:
+              memory: 64Mi
+              cpu: 20m
+            limits:
+              memory: 64Mi
+              cpu: 20m
+      # Schedule onto the game server node pool when running in the same cluster as the inference server.
+      tolerations:
```
Comment (zmerlynn) [inline examples/simple-genai-server/gameserver_npcchat.yaml:60]: In the other examples you commented the node selector hunk out, probably do it here as well.

File: examples/simple-genai-server/main.go [hunk 4]
```diff
@@ -88,6 +90,20 @@ func main() {
 		}
 		simNpc = &snpc
 	}
+	if fid := os.Getenv("FromID"); fid != "" {
```
Comment (zmerlynn) [inline examples/simple-genai-server/main.go:93]: nit: It looks like this pattern is prevalent here already but it's really strange to see environment variables that are not in ALL_UPPER_CASE_WITH_UNDERSCORES. I think we can fix this in a follow-on if you want to leave a TODO, though.

**Review by zmerlynn — CHANGES_REQUESTED** [review]

File: examples/simple-genai-server/gameserver_npcchat.yaml (hunk 3 above)
Comment (igooch) [reply examples/simple-genai-server/gameserver_npcchat.yaml:60]: Done. We'll probably also take this out once the inference server is moved to Autopilot too.

File: examples/simple-genai-server/main.go (hunk 4 above)
Comment (igooch) [reply examples/simple-genai-server/main.go:93]: QQ: should the actual env variables also be ALL_UPPER_CASE_? I.e. should it be `PORT := flag.String("port"...` or `PORT := flag.String("PORT"...`

Comment (zmerlynn) [reply examples/simple-genai-server/main.go:93]: * The flag name *should* be lower-case-with-dashes: https://github.com/googleforgames/agones/blob/main/cmd/controller/main.go#L58-L83
  * _That said_, you're following a nearer convention: https://github.com/googleforgames/agones/blob/main/examples/simple-game-server/main.go#L41-L50 .. so honestly just keep the flag names. Let's fight that battle another day.
* The Go variable name should follow Go style, which is mixedCase: https://go.dev/doc/effective_go#mixed-caps
* The environment variable should be in UPPER_CASE_WITH_UNDERSCORES: https://google.github.io/styleguide/shellguide.html#constants-and-environment-variable-names

File: examples/simple-genai-server/gameserver_npcchat.yaml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      containerPort: 7654
+      protocol: TCP
+  template:
+    spec:
+      containers:
+        - name: simple-genai-game-server
+          image: us-docker.pkg.dev/agones-images/examples/simple-genai-game-server:0.1
+          # imagePullPolicy: Always  # add for development
+          env:
+            - name: GenAiEndpoint
```
Comment (zmerlynn) [inline examples/simple-genai-server/gameserver_npcchat.yaml:32]: these are environment variables, so you'll need to change them to match the name upper case names.

Note that you could have just used args here instead, up to you.

Comment (igooch) [reply examples/simple-genai-server/gameserver_npcchat.yaml:32]: Thanks - missed that. Updated.

**Review by zmerlynn — APPROVED** [review]

**Pull request merged**